src/Javis.jl

```diff
@@ -272,20 +280,48 @@ function render(
         tempdirectory = mktempdir()
     end
 
+    # Create a template useful in case of postprocessing
+    frame_template = _convert_and_rescale_if_needed(
+        get_javis_frame(video, objects, first(frames); layers = layers),
+        rescale_factor,
+    )
+
+    # Check if postprocess_frames_flow is ill defined
+    original_n_frames = length(frames)
+    frames = postprocess_frames_flow(frames)
+    if !issubset(unique(frames), 1:original_n_frames)
+        error(
+            "postprocess_frames_flow should return a vector of frame indices contained in the original number of frames",
+        )
+    end
+
+    # Memoization container for postprocessing
+    frames_memory = Dict{Int,Matrix{RGB{N0f8}}}()
+
+    # helps to check if the video is already being rendered in mp4 case
+    started_encoding_video = false
+
     filecounter = 1
     @showprogress 1 "Rendering frames..." for frame in frames
-        frame_image = convert.(RGB, get_javis_frame(video, objects, frame; layers = layers))
-        # rescale the frame for faster rendering if the rescale_factor is not 1
-        if !isone(rescale_factor)
-            new_size = trunc.(Int, size(frame_image) .* rescale_factor)
-            frame_image = imresize(frame_image, new_size)
-        end
+        frame_image = _postprocess(
+            video,
+            objects,
+            frame,
+            layers = layers,
+            frames_memory = frames_memory,
+            postprocess_frame = postprocess_frame,
+            frame_template = frame_template,
+            filecounter = filecounter,
+            frames = frames,
+            rescale_factor = rescale_factor,
+        )
 
         if !isempty(tempdirectory)
             Images.save("$(tempdirectory)/$(lpad(filecounter, 10, "0")).png", frame_image)
         end
         if render_mp4
-            if frame == first(frames)
+            if frame == first(frames) & !started_encoding_video
```

Comment:
here also the short circuit `&&`

Reply:
I'm unsure why we check for both here actually. Can't we just check it with `if !started_encoding_video` ?

Reply:
Also just for good measure maybe add a test case where the first frame is not `1` which should make a difference.

Reply:
I will add the test! The reason for the check is that if `postprocess_frames_flow` repeats the first frame then the encoding starts twice which I though would be bad, though I don't know VideoIO very much (at all to be fair xD)

Reply:
Yeah only checking for the first frame will fail bit only checking whether we initialized rendering should be enough, or not?

Reply:
For sure I am not understanding smth, but if after `postprocessing_frames_flow` the vector `frames` is [1,2,1] for example would't it start rendering a second time at the third step without the second check?

Reply:
I would only use the second check 🤣

Reply:
ooooooh got I make it now xD